server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java

```diff
@@ -247,11 +235,17 @@ public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode n
                 } else {
                     ret.add(decision);
                 }
-            } else if (decision != Decision.ALWAYS
-                        && (allocation.getDebugMode() != EXCLUDE_YES_DECISIONS || decision.type() != Decision.Type.YES)) {
-                ret.add(decision);
+            } else {
+                addDebugDecision(ret, decision, allocation);
             }
         }
         return ret;
     }
+
+    private void addDebugDecision(Decision.Multi ret, Decision decision, RoutingAllocation allocation) {
+        if (decision != Decision.ALWAYS
+            && (allocation.getDebugMode() == RoutingAllocation.DebugMode.ON || decision.type() != Decision.Type.YES)) {
```

Comment:
perhaps add a comment explaining this condition. I had to rewrite it in my head to `debugMode != ON => type != YES`